addons/mail/models/mail_thread.py

```diff
@@ -1468,25 +1474,62 @@ def _message_parse_extract_bounce(self, email_message, message_dict):
             if bounced_email:
                 bounced_partner = self.env['res.partner'].sudo().search([('email_normalized', '=', bounced_email)])
 
-        bounced_msg_id = False
+        bounced_msg_ids = False
         bounced_message = self.env['mail.message'].sudo()
         if email_part:
             if email_part.get_content_type() == 'text/rfc822-headers':
                 # Convert the message body into a message itself
                 email_payload = message_from_string(email_part.get_content(), policy=email.policy.SMTP)
             else:
                 email_payload = email_part.get_payload()[0]
-            bounced_msg_id = tools.mail_header_msgid_re.findall(tools.decode_message_header(email_payload, 'Message-Id'))
-            if bounced_msg_id:
-                bounced_message = self.env['mail.message'].sudo().search([('message_id', 'in', bounced_msg_id)])
+            bounced_message, bounced_msg_ids = self._get_bounced_message_data(email_payload, message_dict)
+
+        if bounced_message and not bounced_partner and len(bounced_message.notification_ids.res_partner_id) == 1:
+            # if the original recipient was not found,
+            # try to find the recipient based on parent <mail.message> notification
+            bounced_partner = bounced_message.notification_ids.res_partner_id[0]
+            bounced_email = bounced_partner.email
 
         return {
             'bounced_email': bounced_email,
             'bounced_partner': bounced_partner,
-            'bounced_msg_id': bounced_msg_id,
+            'bounced_msg_ids': bounced_msg_ids,
             'bounced_message': bounced_message,
         }
 
+    def _message_parse_is_bounce(self, message, message_dict):
```

Comment:
Answering here: I like this new method, allows to centralize bounce detection.
(It duplicates a bit the email_to_localparts thingy but not a huge deal).